client/finality-grandpa/src/environment.rs

```diff
@@ -1179,13 +1183,29 @@ pub(crate) fn finalize_block<BE, Block, Client>(
 						commit,
 					)?;
 
-					Some(justification.encode())
+					Some(justification)
 				} else {
 					None
 				}
 			},
 		};
 
+		// Notify any registered listeners in case we have a justification
+		if let Some(ref justification) = justification {
+			match client.header(BlockId::Hash(hash)) {
+				Ok(Some(header)) => {
+					let notification = (header, justification.clone());
+					if let Some(sender) = justification_sender {
```

Comment:
I think this should be checked earlier, so that we don't even request `client.header` if `justification_sender.is_none()`.

Reply:
I don't have a strict opinion on this but it could also be done from the RPC side. You only need to send the justification to the RPC side and it can then fetch the header for the given finalized hash (`justification.commit.target_hash`). There's also the possibility that we don't need the header at all as I commented somewhere else :P